### PR TITLE
Feature/box office filters

### PR DESCRIPTION
**Pull request by walkerkit**
Think this is good to go, could do some comments - they might be a bit, how you say ... unmanageable

### CODE REVIEW

**Comment by walkerkit** [comment]
well, that's tragic

File: README.md [hunk 1]
```diff
@@ -94,13 +94,13 @@ make test-v
...
 
 ```
-docker-compose run --rm api pytest --cov uobtheatre -vv -s uobtheatre/bookings/test/test_views.py
+make test path=uobtheatre/bookings/test/test_views.py
 ```
 
 and to run a single test
 
 ```
-docker-compose run --rm api pytest --cov uobtheatre -vv -s uobtheatre/bookings/test/test_views.py -k 'test_name'
```
Comment (JElgar) [inline README.md:103]: ❤️

File: uobtheatre/bookings/models.py [hunk 2]
```diff
@@ -74,6 +79,66 @@ class Meta:
...
+
+    def annotate_checked_in(self):
+        return self.annotate(checked_in=BoolAnd("tickets__checked_in"))
+
+    def annotate_not_checked_in(self):
+        return self.annotate(count=models.Count("tickets")).annotate(
+            checked_in_count=models.Count(
+                Case(When(tickets__checked_in=True, then=Value(1)))
+            )
+        )
```
Comment (JElgar) [inline uobtheatre/bookings/models.py:93]: Does this do what it says on the tin `annotate_not_checked_in`?

File: uobtheatre/bookings/models.py [hunk 3]
```diff
@@ -74,6 +79,66 @@ class Meta:
...
+
+    def annotate_checked_in_proportion(self):
+        return self.annotate_not_checked_in().annotate(
+            proportion=Case(
+                When(Q(count=0), then=Cast(0, FloatField())),
+                default=Cast(F("checked_in_count"), FloatField())
+                / Cast(F("count"), FloatField()),
+            )
```
Comment (JElgar) [inline uobtheatre/bookings/models.py:101]: Very cool

File: uobtheatre/bookings/models.py [hunk 4]
```diff
@@ -74,6 +79,66 @@ class Meta:
...
+        )
+
+    def checked_in(self, bool_val=True):
```
Comment (JElgar) [inline uobtheatre/bookings/models.py:104]: Can you type function definitions aswell plz?

File: uobtheatre/bookings/models.py [hunk 5]
```diff
@@ -74,6 +79,66 @@ class Meta:
...
+        if bool_val:
+            query_set = self.annotate_checked_in().filter(checked_in=True)
+        else:
+            query_set = self.annotate_not_checked_in().filter(
+                checked_in_count__lt=F("count")
+            )
+
+        return query_set
+
+    def active(self, bool_val=True):
```
Comment (JElgar) [inline uobtheatre/bookings/models.py:124]: Im not a huge fan of just using `bool_val` can we do something like:

```
Booking.objects.filter_active(active=True)
```

File: uobtheatre/bookings/schema.py [hunk 6]
```diff
@@ -130,6 +130,51 @@ class Meta:
...
+            query_set (QuerySet): The Queryset which is being filtered.
+            value (str): The choices s(eg 'start')
+
+        Returns:
+            Queryset: The filtered Queryset
+        """
+        if value and "created_at" in value:
+            return query_set.order_by("created_at")
+        if value and "-created_at" in value:
+            return query_set.order_by("-created_at")
```
Comment (JElgar) [inline uobtheatre/bookings/schema.py:168]:
```suggestion
        if value in ["created_at", "-created_at"]:
            return query_set.order_by(value)  
```

File: uobtheatre/bookings/models.py (hunk 5 above)
Comment (JElgar) [reply uobtheatre/bookings/models.py:124]: I lie I just looked at the tests and that's very nice

**Review by JElgar — APPROVED** [review]
This is beautiful stuff thank you @walkerkit!

**Comment by JElgar** [comment]
I've approved changes with a few small comments which would be nice to have. I also agree adding a few more docstrings to the queryset class would be noice.

To get the schema ci to pass just run `make schema` locally.

File: uobtheatre/bookings/models.py (hunk 2 above)
Comment (walkerkit) [reply uobtheatre/bookings/models.py:93]: um... no

File: uobtheatre/bookings/models.py (hunk 4 above)
Comment (walkerkit) [reply uobtheatre/bookings/models.py:104]: yup

**Comment by walkerkit** [comment]
don't really know why this has started to fail

**Comment by walkerkit** [comment]
@JElgar I don't know why this is failing on CI - it's working locally!

**Comment by walkerkit** [comment]
good to merge

File: uobtheatre/bookings/models.py [hunk 7]
```diff
@@ -74,6 +79,70 @@ class Meta:
...
+            QuerySet: the filtered queryset
+        """
+
+        if bool_val:
+            query_set = self.annotate_checked_in().filter(checked_in=True)
+        else:
+            query_set = self.annotate_checked_in_count().filter(
+                checked_in_count__lt=F("count")
+            )
+        print(query_set)
```
Comment (JElgar) [inline uobtheatre/bookings/models.py:125]:
```suggestion
```

**Pull request merged**